### PR TITLE
Make SRI hashes configurable, no longer hardcoded in templates

### LINKED ISSUE #1365: Injecting the SRI via template variables instead of hardcoding them into the templates
This is a tangential idea based off a [recent discussion](https://github.com/orgs/PrivateBin/discussions/1363#discussioncomment-9885278)

## The problem

For those folks that create their custom or modified templates, a task they have to do upon every upgrade is to update the SRI hashes in the script tags. Of course they may also have to add any new HTML elements we might have added, but to be fair, that is rare.

## The solution

The SRI-hashes are the same for all templates, and could be injected into them from a single source of truth, for example from default configuration values. That way they could still be overridden in the configuration file, if really needed (=when they modify the JS themselves), but otherwise they would always be up to date upon the upgrade.

Can anyone think of any risks that this would introduce?

## Alternatives

We decide the current solution is adequate and we leave everything as it is.

## Additional context

This is a bit of a papercut-issue, annoying but not blocking. It may lead to admins avoiding updates or avoiding to customize the templates, which kinda defeats the point of the feature (templates).

### PR DESCRIPTION
**Pull request by elrido**
This PR fixes #1365

## Changes
* should make upgrades easier for those using custom templates
* SRI hashes are now stored as default values in lib/Configuration.php
* if the JS files got customized, the default SRI hashes can be replaced using the conf.php file, added commented section in conf.sample.php
* If no hash is found in the SRI array, no integrity attribute is added - this is a bit of a risk in that one could forget to update the file name in lib/Configuration.php when upgrading the version of a JS file, this would no longer cause an error during manual testing, but also no longer protect the JS file's integrity

Should we make failures of the SRI lookup hard errors? It would make the application more fragile, but also prevented silent errors updating the hashes, i.e. in unit tests. Or we could log an error message but still render the template? Or maybe add a unit test that detects missing SRI hashes for js files (hard failure in unit test, soft failure on admin misconfiguration)?

What doesn't change is that we can regenerate the hashes using phpunit.

### CODE REVIEW

File: lib/View.php [hunk 1]
```diff
@@ -57,4 +57,22 @@ public function draw($template)
...
+     * @param  string $file
+     * @param  bool   $async should it execute ASAP or only after HTML got parsed
+     */
+    private function _scriptTag($file, $async = true)
+    {
+        $sri = array_key_exists($file, $this->_variables['SRI']) ?
+            ' integrity="' . $this->_variables['SRI'][$file] . '"' : '';
+        $suffix = preg_match('#\d.js$#', $file) == 0 ?
+            '?' . rawurlencode($this->_variables['VERSION']) : '';
+        echo '<script ', $async ? 'async' : 'defer',
```
Comment (rugk) [inline lib/View.php:74]: Is it again my PHP knowledge lacking or why do you return the string comma-separated and not concatenated with `.`?

Note the PHPdoc here also does not document the return value, so I'm a little puzzled…

File: lib/View.php [hunk 2]
```diff
@@ -57,4 +57,22 @@ public function draw($template)
...
+     *
+     * @access private
+     * @param  string $file
+     * @param  bool   $async should it execute ASAP or only after HTML got parsed
+     */
+    private function _scriptTag($file, $async = true)
+    {
+        $sri = array_key_exists($file, $this->_variables['SRI']) ?
+            ' integrity="' . $this->_variables['SRI'][$file] . '"' : '';
+        $suffix = preg_match('#\d.js$#', $file) == 0 ?
```
Comment (rugk) [inline lib/View.php:72]: RegEx always could have a negative performance effect (in theory)… (Note RegEx is usually the slowest thing to implement pattern matching due to it's potential complexity) as this is executed inline for each paste access/template…

Could we maybe "outsource" this cost to some "config validation" that is only run once ("at startup" I would say if this were not a PHP application, because obviously it does not have this hmm).

Also code-readability-wise I don't quite get:
* what are these hash signs for? … ahhhh looking at PHP dic it's sed(-like) syntax for allowing slashes okay, fair enough, just me not knowing PHP commands here
* then, however, it matching `.js` ending files, okay, but why the `\d`, is it really relevant to only match JS files that end with ~~a~~ one digit aka number? Why not all files? 
* also, is not the logic inversed here: When finds 0 matches aka does not match a JS file, then append the version? Maybe inverse the if here, so the "real" "true" condition comes first?
* note one may want to mention the reason (I'm fine with a commend) for this: adding a ~~cache breaker~~ [cache buster](https://stackoverflow.com/questions/9692665/cache-busting-via-params) (that's apparently the correct term) – or just generally extract it into a function like "getJsSuffix"?

**Review by rugk — COMMENTED** [review]
Yeah i wondered how you'd make the "how do I customize the hash" thing and obviously config options default's are an elegant and simple solution! Thanks!

This basically boils down to update the configuration file (as usually done and needed if one wants to use non-default new config stuff or in case should we change a config key or so), which is fine and usually does not need to be done.

One thing to note:
* obviously, this makes the template adjustments much more complex if one chooses to somehow customize the `script` tag in itself, aka other attributes mostly: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script
  * [`fetchpriority`]([fetchpriority](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#fetchpriority)) could be such a case (also for us maybe?)
  * an idea to make this more adustable and at the same time improve the API by removing bootlean parameter (generally not a good sign for function headers), would be to change the API to `private function _scriptTag($file, $attributes = 'async="false"')` (spaces around it should probably be added automatically and here I just use a default to omit the async parameter)

**Comment by rugk** [comment]
> Should we make failures of the SRI lookup hard errors?

That's fine IMHO? So technically, this would mean making `_scriptTag` error when an SRI hash cannot be found, does it?
In such a case, this is either a programming error (here in the source) or a fundamental configuration error, so IMHO this is fine.
A unit test would in any case be good in case some contributor (or me :wink: ) to submit a change without testing it manually. IMHO though, we can always fail hard?

> this would no longer cause an error during manual testing

Ah though you have a point here. This can indeed ease development. We don't have a development mode yet, do we? Otherwise I would have said, if that is enabled yeah, keep the current behavior is a good idea, so people don't need to update stuff. (A development config file would then be a `privatebin.js = ""` or so, containing an empty string? – so unit test changes the defaults only?) Though, yet again, this would have the disadvantage that in development nobody tests the actual page with SRI hashes - if one forgets to run the unit tests that again could have problematic consequences?
I dunno, I am unsure about this one…

Ah and BTW currently tests are still failing…

File: lib/View.php (hunk 1 above)
Comment (elrido) [reply lib/View.php:74]: There is no return value, the string is directly echo-ed. Since this is a private method, we can't (easily) unit test it, so I felt no need to add the extra return and add echos before every call to it in the templates.

This then allows a further micro-optimization: When using an echo keyword (like return, this is not a function) we can avoid the creation of an extra string in memory to catenate all the pieces into, but just pass all pieces as arguments to echo, hence the commas. See some other echos used in the templates, for example:
https://github.com/PrivateBin/PrivateBin/blob/3cba170f3255de21bbebb77f6c565519ef33e8c1/tpl/bootstrap5.php#L165

**Comment by elrido** [comment]
>     * obviously, this makes the template adjustments much more complex if one chooses to somehow customize the `script` tag in itself, aka other attributes mostly: 

Not really, if someone wanted to customize a script tag, they could just write the script tag by directly and not use this function - they could opt to use the SRI hash via a lookup in the $SRI variable or omit it or hardcode it.

>       * an idea to make this more adjustable and at the same time improve the API by removing the boolean parameter (generally not a good sign for function headers), would be to change the API to `private function _scriptTag($file, $attributes = 'async="false"')` (spaces around it should probably be added automatically and here I just use a default to omit the async parameter)

Let's consider doing that, but using an array, if and when we ever need a different attribute. For now it would just needlessly introduce complexity to the code.

**Comment by rugk** [comment]
Hmm I find that not really more complex than the boolean flag param, and it increases readability. Let's say you want to custimize a template and did not know anything about PrivateBin's code base, would you know what these `true` and `false` randomly speckled in there mean? I doubt so… you likely use no real IDE but maybe vim or nano or so on a server (okay, vim fans will kill me for this statement haha, but anyway), so you will not want or look up the function definition.


```php
		<?php $this->_scriptTag('js/privatebin.js', false); ?>
		<?php $this->_scriptTag('js/privatebin.js', 'async'); ?>
		<?php $this->_scriptTag('js/privatebin.js', ['async']); ?>		
```

Basically, we create our own micro-syntax for the JS script tag, so that's why I'd say the first two versions are more readable and easier to understand, while being flexible for more modifications, if needed. Personally, I said just a string, because I think it would be easier to implement, but I'd be fine with an array, too.
Note template authors need to adjust this, if we later change it… so if we give them an API now, we should rather keep it stable (we never considered it as such, but it's a "breaking change" basically, if we change the method then later).

File: tpl/bootstrap5.php [hunk 3]
```diff
@@ -25,39 +25,39 @@
...
+		<?php $this->_scriptTag('js/showdown-2.1.0.js'); ?>
 <?php
 endif;
 ?>
-		<script type="text/javascript" data-cfasync="false" src="js/purify-3.1.3.js" integrity="sha512-t/FKG/ucQVMWTWVouSMABSEx1r+uSyAI9eNDq0KEr9mPhkgxpJztHI/E72JIpv/+VwPs/Q4husxj14TE9Ps/wg==" crossorigin="anonymous"></script>
-		<script type="text/javascript" data-cfasync="false" src="js/legacy.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-LYos+qXHIRqFf5ZPNphvtTB0cgzHUizu2wwcOwcwz/VIpRv9lpcBgPYz4uq6jx0INwCAj6Fbnl5HoKiLufS2jg==" crossorigin="anonymous"></script>
-		<script type="text/javascript" data-cfasync="false" src="js/privatebin.js?<?php echo rawurlencode($VERSION); ?>" integrity="sha512-cbmXvtZ/5gZPFjQDzP3IEhUAIhFPAoM31gw2kRYCT5xOh8wv9gXeDqI/t798luRW1xdC4gaYodjEFCzrsZR4mA==" crossorigin="anonymous"></script>
+		<?php $this->_scriptTag('js/purify-3.1.3.js'); ?>
+		<?php $this->_scriptTag('js/legacy.js'); ?>
+		<?php $this->_scriptTag('js/privatebin.js', false); ?>
```
[the remark this answers is not in the document]
Comment (elrido) [reply tpl/bootstrap5.php:60]: I played with the settings and discovered that most scripts can be loaded async, except that bootstrap 3 requires jquery to already be loaded, so there we load them both deferred, but bootstrap 5 doesn't, so it can remain async. PB itself always needs to start deferred, since it launches as soon as the HTML is loaded, but it itself require jquery, too. So we do now load most scripts async and are explicit about the deferred start. Before we defaulted to sequential loading and start once all scripts and HTML are fully loaded, except the async QR code one.

**Comment by elrido** [comment]
Ah, I didn't get that you intended the parameter to just be a string and not an array. Ok, that is less complex then even a boolean, while adding some flexibility, so I'll take that change into account.

File: lib/View.php (hunk 2 above)
Comment (elrido) [reply lib/View.php]: 4b6c835 should address this

**Comment by elrido** [comment]
Apart from the attributes string parameter, the unit test integrated SRI update mechanism will now complain if a hash is found in that lib/Configuration.php when there is no matching js file as well as if there are more js files found then are present in the configuration (but common.js used by mocha is ignored). That way no runtime check needs to get performed, though an admin still has the option to disable the SRI hashes by configuring an incomplete list in their conf.php. We would have to assume that to get done on purpose.

**Comment by rugk** [comment]
> an admin still has the option to disable the SRI hashes by configuring an incomplete list in their conf.php. We would have to assume that to get done on purpose.

Uhm, but what if we introduce a new JS file and obviously admins may or may not update their list?

Okay one solution would be: merge that list with the defaults we provide, so this cannot happen?

**Review by rugk — COMMENTED** [review]
Ah just saw in `tst/Bootstrap.php` such a check is implemented, but in a production system with such a config, an admin could still fall in that trap and loose SRI as a security feature… kinda.
IMHO, we should do our best to prevent this.

Like in the worst case, if no hash is found we could add an invalid one, which will cause errors and thus cause admins to think about what they have customized. (To allow the use case of removing the hash, one could introduce a special marker/config like `privatebin.js = insecure-withoutsri` or so. This would make the consequence obvious and the decision a deliberate one.)
Just as an idea.

**Comment by elrido** [comment]
While I find that merging this list overcomplicates things and introduces further risks, I'll see if we can merge the lists easily, without causing other side effects on other arrays (for example, we do not want the expiration lists to get merged between defaults and custom configuration).

So it would only add the default value if the configured list doesn't contain the key. ATM it fails without breaking things and an admin may want to turn off SRIs. We would no longer allow for that with the merged list thing and force them to create valid list entries for existing files (that they modified). With this merged list, they could still introduce new JS files into a custom template without hashes, but then they can also simply replace the function call with a hardcoded script tag without SRI, so I don't think that is something we can or need to guard for. Someone that does want to remove the SRIs on purpose will find many ways to do so.

File: tst/ConfigurationTest.php [hunk 4]
```diff
@@ -120,9 +120,15 @@ public function testHandleMissingSubKeys()
         unset($options['expire_options']['1week']);
         unset($options['expire_options']['1year']);
         unset($options['expire_options']['never']);
+        $sri_key                         = array_key_first($options['sri']);
+        $valid_sri                       = $options['sri'][$sri_key];
+        $options['sri'][$sri_key]        = ''; // empty string should get replaced with default
```
Comment (rugk) [inline tst/ConfigurationTest.php:125]: Hmm maybe I do understand the code wrong, but where is the assert for this? Your change only adds prepare steps, it does not change what is actually asserted at the end.


Ahhhh the `$options`  at the end is modified again. Well for readability I would instead clone it and name the variable something like `$expectedOptions`. (Also some newline before would not hurt).
However, this means `some invalid SRI hash` is filtered out? That's cool, but I thought you did not want to introduce validation? Ah okay, just because it's non-empty.
Then maybe name it `invalid SRI hash should be kept` to make that clear.

**Review by rugk — APPROVED** [review]
Yeah only a small nit.

File: tst/ConfigurationTest.php (hunk 4 above)
Comment (elrido) [reply tst/ConfigurationTest.php:125]: Correct interpretation, yes. The empty key gets replaced because of PHP's ini parser ignoring it and the incorrect one is preserved because we don't validate it (because that would require constantly regenerating SRIs on every request, in case the files got changed, we could then cache that to reduce the load, but then have even more complex code for that caching, etc.)

I had added the comment regarding restoring the expected values for that clarification. I think this dragged out way too long already and I'll merge it as is. If you do care about the naming of the vars, feel free to raise another PR for that. THX
